### PR TITLE
Feature/nextjs blog list and blog view

### PR DESCRIPTION
**Pull request by jordanch**
outline blog and post 

todo: pagination 
todo: subscribe method
todo: style

### CODE REVIEW

File: examples/nextjs-blog/pages/blog/list.js [hunk 1]
```diff
@@ -1,20 +1,58 @@
...
 
 const Posts = function(props) {
-  return Object.values(props.posts).map(post => (
-    <h3>
-      <Link route={`/blog/${post.slug}`}>{post.title}</Link>
-    </h3>
-  ))
+  const { posts } = props
+
+  if (posts && posts.length > 0) {
```
Comment (gitdubz) [inline examples/nextjs-blog/pages/blog/list.js]: no need to check > 0 🍡

File: examples/nextjs-blog/pages/blog/list.js [hunk 2]
```diff
@@ -1,20 +1,58 @@
...
+      )
+    })
+  }
+
+  return 'No blog posts!'
 }
 
-Posts.getInitialProps = async ({ query }) => {
+Posts.getInitialProps = async () => {
+  const posts = await app.content.get({ schemaKey: 'blogPost', populate: true })
```
Comment (gitdubz) [inline examples/nextjs-blog/pages/blog/list.js]: can we add the sdk filter for this and not filter client side as an example?

Comment (gitdubz) [reply examples/nextjs-blog/pages/blog/list.js]: @jperasmus how do we do a where lookup and orderBy for CF?

Might be something to add to the docs

`
app.content.get({ ..., where: ["status == 'published'"], orderBy: [{ field: 'date',  order: 'desc'}] })
`

File: examples/nextjs-blog/pages/blog/list.js [hunk 3]
```diff
@@ -1,20 +1,58 @@
...
+          const dateB = new Date(postB.date)
+
+          if (dateA < dateB) return -1
+          if (dateA > dateB) return 1
+          return 0
+        })
+    )
+  )
+
+  console.log(filteredPosts)
```
Comment (gitdubz) [inline examples/nextjs-blog/pages/blog/list.js]: consolas

File: examples/nextjs-blog/pages/blog/post.js [hunk 4]
```diff
@@ -1,12 +1,58 @@
...
+        <img src={imageURL} alt={getImageAlt(post)} />
+      </div>
+    )
+  }
+
+  return 'No post data!'
 }
 
 Post.getInitialProps = async ({ query }) => {
   console.log('SLUG', query.slug)
```
Comment (gitdubz) [inline examples/nextjs-blog/pages/blog/post.js]: consolas

File: examples/nextjs-blog/utils/post/post.util.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { flamelinkApp as app } from '../flamelink'
+
+export function getPostsWithMedia(posts = []) {
+  return (
+    posts &&
+    posts.map(async post => {
+      const [image] = post.image
```
Comment (gitdubz) [inline examples/nextjs-blog/utils/post/post.util.js]: @jperasmus I am assuming this won't be needed after the tiny little url issue (that is not so easy) is resolved?

File: examples/nextjs-blog/pages/blog/list.js (hunk 2 above)
Comment (jordanch) [reply examples/nextjs-blog/pages/blog/list.js]: ah ofc, dankie

File: examples/nextjs-blog/pages/blog/list.js [hunk 6]
```diff
@@ -19,7 +19,11 @@ const Posts = function(props) {
 
       return (
         <div key={docId}>
-          <Link route={`/blog/${slug}`}>{`${title} - ${displayName}`}</Link>
+          <Link route={`/blog/${slug}`}>
+            {/* href gets added to <a> by <Link> */}
+            {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
+            <a>{`${title} - ${displayName}`}</a>
```
Comment (jperasmus) [inline examples/nextjs-blog/pages/blog/list.js]: Is the Link not adding an `<a>` tag already? Why add another one?

File: examples/nextjs-blog/pages/blog/post.js [hunk 7]
```diff
@@ -1,38 +1,89 @@
...
+    }
   }
 
-  return 'No post data!'
+  componentDidMount() {
+    // todo: this isn't so lekker because the path may change in future. is it
+    // possible to share the `query.slug` value from the server's execution with
+    // the client?
+    const { pathname } = location
+    const slug = pathname.replace('/blog/', '')
```
Comment (jperasmus) [inline examples/nextjs-blog/pages/blog/post.js]: maybe split the `pathname` by `/` and pop off the last item in the array as the `slug`?

File: examples/nextjs-blog/pages/blog/post.js [hunk 8]
```diff
@@ -1,38 +1,89 @@
...
+      callback: async (error, response) => {
+        if (error) {
+          throw new Error(
+            'Something went wrong while retrieving all the content. Details:',
+            error
+          )
+        }
+
+        const [post] = await Promise.all(
+          getPostsWithMedia(Object.values(response || {}))
```
Comment (jperasmus) [inline examples/nextjs-blog/pages/blog/post.js]: The latest version of the SDK should include the `url` property for files when you `populate` them. I'll make sure the latest SDK is properly published today. I published last night, but something went wrong half-way through and only half of the packages got published.

File: examples/nextjs-blog/pages/blog/list.js (hunk 6 above)
Comment (gitdubz) [reply examples/nextjs-blog/pages/blog/list.js]: the nextjs one does not do this, if you add a `<a>` it adds the href and then when you hover it is accessible (shows the url in the bottom of you screen like a normal link)

File: examples/nextjs-blog/pages/blog/list.js (hunk 2 above)
Comment (gitdubz) [reply examples/nextjs-blog/pages/blog/list.js]: we have the filters
```
app.content.get({ 
  ..., 
  filters: [ ["status", "==" "published"] ], 
  orderBy: [{ field: 'date', order: 'desc'}] 
})
```

File: examples/nextjs-blog/utils/post/post.util.js (hunk 5 above)
Comment (jperasmus) [reply examples/nextjs-blog/utils/post/post.util.js]: You got jokes, nê? Should be fixed with this PR: https://github.com/flamelink/examples/pull/8

File: examples/nextjs-blog/pages/blog/list.js (hunk 6 above)
Comment (jperasmus) [reply examples/nextjs-blog/pages/blog/list.js]: that is very weird, but thanks for the info. Now the comment above makes more sense

File: examples/nextjs-blog/pages/blog/list.js (hunk 2 above)
Comment (jperasmus) [reply examples/nextjs-blog/pages/blog/list.js]: yup, lekker, you've got it @gitdubz

File: examples/nextjs-blog/pages/blog/post.js (hunk 8 above)
Comment (jperasmus) [reply examples/nextjs-blog/pages/blog/post.js]: PR #8 should resolve this, so then you can remove the manual query

File: examples/nextjs-blog/pages/blog/post.js [hunk 9]
```diff
@@ -1,38 +1,89 @@
...
-        <img src={imageURL} alt={getImageAlt(post)} />
-      </div>
-    )
+import {
+  getPostsWithMedia,
+  getImageAlt,
+  getDateString,
+} from '../../utils/post/post.util'
+
+class Post extends PureComponent {
```
[the remark this answers is not in the document]
Comment (jordanch) [reply examples/nextjs-blog/pages/blog/post.js]: I hear you, in favouring speed I'll stick with classes.

[the remark this answers is not in the document]
Comment (jperasmus) [reply examples/nextjs-blog/pages/blog/post.js]: Not sure I follow. Are you referring to speed of development or the code itself?

File: examples/nextjs-blog/pages/blog/post.js [hunk 10]
```diff
@@ -17,29 +14,26 @@ class Post extends PureComponent {
...
-  componentDidMount() {
-    // todo: this isn't so lekker because the path may change in future. is it
-    // possible to share the `query.slug` value from the server's execution with
-    // the client?
+  async componentDidMount() {
     const { pathname } = location
-    const slug = pathname.replace('/blog/', '')
+    const slug = pathname.split('/').pop()
 
+    // todo: SDK bug - subscribing using an event type doesn't work
```
Comment (jperasmus) [inline examples/nextjs-blog/pages/blog/post.js]: Hey @jordanch what is currently happening?

File: examples/nextjs-blog/pages/blog/post.js [hunk 11]
```diff
@@ -17,29 +14,26 @@ class Post extends PureComponent {
...
     this._subscription = app.content.subscribe({
       schemaKey: 'blogPost',
-      changeType: 'modified',
-      filters: [['slug', '==', slug]],
-      populate: true,
-      callback: async (error, response) => {
+      // changeType: 'modified',
+      filters: [['slug', '==', slug], ['_fl_meta_.locale', '==', 'en-US']],
```
Comment (jperasmus) [inline examples/nextjs-blog/pages/blog/post.js]: You don't need to filter on the locale here, the SDK does that for you.

**Comment by jperasmus** [comment]
Resolved merge conflicts for `package.json` file

**Comment by jperasmus** [comment]
Pushed a couple of styling updates for the Post pages. Still very much WIP

File: examples/nextjs-blog/pages/_app.js [hunk 12]
```diff
@@ -48,6 +48,7 @@ class MyApp extends App {
             --footer-bg-color: #161616;
             --footer-text-color: #fff;
             --container-wide: 85rem;
+            --container-medium: 56rem;
```
Comment (gitdubz) [inline examples/nextjs-blog/pages/_app.js]: if we are using material-ui we can just use the grids they provide?

File: examples/nextjs-blog/pages/blog/post.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (jordanch) [reply examples/nextjs-blog/pages/blog/post.js]: Speed of development; I'm uncertain but I thought the goal was to finish this last week.

File: examples/nextjs-blog/pages/blog/post.js (hunk 10 above)
Comment (jordanch) [reply examples/nextjs-blog/pages/blog/post.js]: @jperasmus Bug logged against SDK Epic. Inline here for convenience: 

JS SDK v1 > Content > subscribe > subscribe with changeType doesn't call callback on change

https://responsivestudio.atlassian.net/browse/FL-1514

When adding the changeType option, the callback doesn't fire on document event.

File: examples/nextjs-blog/pages/blog/post.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (jperasmus) [reply examples/nextjs-blog/pages/blog/post.js]: Yeah, no problem, just wanted to understand what you meant. 👍 We definitely wanted to get it done last week, but the 3-day week with the load-shedding really put a damper on our efforts.

File: examples/nextjs-blog/pages/blog/post.js (hunk 10 above)
Comment (jperasmus) [reply examples/nextjs-blog/pages/blog/post.js]: Thanks for logging, I'll check it out.

**Comment by jordanch** [comment]
Can we please get this into the main branch and open other PRs - this one is getting hairy

**Pull request merged**